### ISSUE TITLE
Node is stuck in infinite 'try_insert_validated_block' loop (+ DB breakage)

### ISSUE THREAD

**Issue by Gaon3** [paradigmxyz/reth#3087]
### Describe the bug

Hello,
At some point, the node got stuck at trying to insert a validated block, with the block number decreasing by one every time. 
Then for a while it tried to insert the block 17443735 (block number stopped decreasing). 

After 
```
2023-06-09T19:13:45.523002Z ERROR make_canonical{block_hash=0x9bda19acb197ae5332e16bf7bd748d0fafa54fbaf7a8443eadf3d0807598541a}: blockchain_tree: Block hash not found in block indices block_hash=0x9bda19
acb197ae5332e16bf7bd748d0fafa54fbaf7a8443eadf3d0807598541a                                                                                                                                                 
2023-06-09T19:13:48.587910Z  INFO try_insert_validated_block{block=(17444669, 0x950ca49fcc79646aae09f4b16e530b6b769c4017217030ff39dde1cbb9a7c6db)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444666, 0x5bd6a3723fa8fdd70f6b09ddc2bdbc36cb5c40657ad0de27d7972c59c230d67d) })                                                                                                                       
2023-06-09T19:13:48.677349Z ERROR make_canonical{block_hash=0x950ca49fcc79646aae09f4b16e530b6b769c4017217030ff39dde1cbb9a7c6db}: blockchain_tree: Block hash not found in block indices block_hash=0x950ca4
9fcc79646aae09f4b16e530b6b769c4017217030ff39dde1cbb9a7c6db                                                                                                                                                 
2023-06-09T19:13:48.677571Z  WARN consensus::engine: Error canonicalizing the head hash error=Execution(BlockHashNotFoundInChain { block_hash: 0x950ca49fcc79646aae09f4b16e530b6b769c4017217030ff39dde1cbb9
a7c6db }) state=ForkchoiceState { head_block_hash: 0x950ca49fcc79646aae09f4b16e530b6b769c4017217030ff39dde1cbb9a7c6db, safe_block_hash: 0xcef73270a313c9b1119115ac8b7d820ef2012f2ea3c8f88665dab09d98465eca,
 finalized_block_hash: 0x2255379e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd0347f1 }
```

It tried to insert starting from 17444666 with the number decreasing every time again.

After pulling main 6ef3e12f (I didn't restart the node directly, I pulled then restarted it) I got this error: 
```
2023-06-09T19:21:19.336539Z ERROR make_canonical{block_hash=0x2255379e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd0347f1}: blockchain_tree: Block hash not found in block indices block_hash=0x225537
9e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd0347f1                                                                                                                                                 
2023-06-09T19:21:19.336560Z  WARN consensus::engine: Error canonicalizing the head hash error=Execution(BlockHashNotFoundInChain { block_hash: 0x2255379e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd
0347f1 }) state=ForkchoiceState { head_block_hash: 0x2255379e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd0347f1, safe_block_hash: 0x2255379e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd0347f1,
 finalized_block_hash: 0x2255379e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd0347f1 }                                                                                                                
2023-06-09T19:21:19.336584Z ERROR make_canonical{block_hash=0x2255379e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd0347f1}: blockchain_tree: Block hash not found in block indices block_hash=0x225537
9e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd0347f1                                                                                                                                                 
2023-06-09T19:21:19.336592Z  WARN consensus::engine: Error canonicalizing the head hash error=Execution(BlockHashNotFoundInChain { block_hash: 0x2255379e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd
0347f1 }) state=ForkchoiceState { head_block_hash: 0x2255379e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd0347f1, safe_block_hash: 0x2255379e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd0347f1,
 finalized_block_hash: 0x2255379e90a2dd147d025f79175c209db9d346826fcda9be86b79df3bd0347f1 }                 
```

Which ends with a DB breakage... 

`2023-06-09T19:22:07.169766Z ERROR execute{stage=Bodies}: sync::pipeline: Stage encountered a fatal error: An internal database error occurred: Database write error code: -30796. stage=Bodies`

### Steps to reproduce

idk

### Node logs

```text
2023-06-09T19:14:07.228469Z  INFO try_insert_validated_block{block=(17444633, 0xc27c25e41deb85189c8f90a3b5b90ebe8a12c94714d2e3f933ded59a6085389d)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444632, 0x7782fdc7f693f8196bc0d1ac601ff685e5cf26f745959bf0c7e0d9924157fc47) })                                                                                                                       
2023-06-09T19:14:07.615008Z  INFO try_insert_validated_block{block=(17444632, 0x7782fdc7f693f8196bc0d1ac601ff685e5cf26f745959bf0c7e0d9924157fc47)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444631, 0xd0270eaf4500f37f124ade22eefafa856ec80604f767cce9d7bc6f01b55e1fd2) })                                                                                                                       
2023-06-09T19:14:07.798041Z  INFO try_insert_validated_block{block=(17444631, 0xd0270eaf4500f37f124ade22eefafa856ec80604f767cce9d7bc6f01b55e1fd2)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444630, 0x83e599dab64379a49399315a708b1412aaee8bda27d5d491e247e3a581f035ce) })                                                                                                                       
2023-06-09T19:14:07.980940Z  INFO try_insert_validated_block{block=(17444630, 0x83e599dab64379a49399315a708b1412aaee8bda27d5d491e247e3a581f035ce)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444629, 0xdf6cc16dbb8db824c665c7191c49d025c43f7e08f1f762439c290329e7eea0d6) })                                                                                                                       
2023-06-09T19:14:08.323602Z  INFO try_insert_validated_block{block=(17444629, 0xdf6cc16dbb8db824c665c7191c49d025c43f7e08f1f762439c290329e7eea0d6)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444628, 0x40e6423684806132a901ec11806e5c91b501b66f1d90d062d2f493ddb1b997cd) })                                                                                                                       
2023-06-09T19:14:08.703067Z  INFO try_insert_validated_block{block=(17444628, 0x40e6423684806132a901ec11806e5c91b501b66f1d90d062d2f493ddb1b997cd)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444627, 0x3a04f9d5ad4e6674724d30f19423943ad45d7c9b4140461f5cde494fbe11b539) })                                                                                                                       
2023-06-09T19:14:08.883154Z  INFO try_insert_validated_block{block=(17444627, 0x3a04f9d5ad4e6674724d30f19423943ad45d7c9b4140461f5cde494fbe11b539)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444626, 0xd5ab18857c379b5a967174f73fdeb43e8e025071630537f8254c9a814fe06946) })                                                                                                                       
2023-06-09T19:14:09.238289Z  INFO try_insert_validated_block{block=(17444626, 0xd5ab18857c379b5a967174f73fdeb43e8e025071630537f8254c9a814fe06946)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444625, 0xd08bad474bd057baeeec43270f11f85b574555895942373f072432928b8ee1fe) })                                                                                                                       
2023-06-09T19:14:09.580762Z  INFO try_insert_validated_block{block=(17444625, 0xd08bad474bd057baeeec43270f11f85b574555895942373f072432928b8ee1fe)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444624, 0x7376a83139423d870dc7037d08a7978837de02108ed9b5819e37c161462f6003) })                                                                                                                       
2023-06-09T19:14:09.922105Z  INFO try_insert_validated_block{block=(17444624, 0x7376a83139423d870dc7037d08a7978837de02108ed9b5819e37c161462f6003)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444623, 0x987c36be1896d17e07389c05d394b0d14eb5048d8a7765bdc9943fdfa214d5c8) })                                                                                                                       
2023-06-09T19:14:10.266936Z  INFO try_insert_validated_block{block=(17444623, 0x987c36be1896d17e07389c05d394b0d14eb5048d8a7765bdc9943fdfa214d5c8)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444622, 0x84461071c7b51ef9af1f39c7c0465726b4fa45a028a7578dd6d7340d9bd30b20) })                                                                                                                       
2023-06-09T19:14:10.447999Z  INFO try_insert_validated_block{block=(17444622, 0x84461071c7b51ef9af1f39c7c0465726b4fa45a028a7578dd6d7340d9bd30b20)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444621, 0xda4fb7fe09ad977c75bf193916a4c067f72153fdc471ada7cc2655165566a13f) })                                                                                                                       
2023-06-09T19:14:10.793730Z  INFO try_insert_validated_block{block=(17444621, 0xda4fb7fe09ad977c75bf193916a4c067f72153fdc471ada7cc2655165566a13f)}: blockchain_tree: return=Ok(Disconnected { missing_paren
t: (17444620, 0x21798c986e7da5bc5503739c5ec25515987086c21c7e8680de1d82916eac3f18) })
```

### Platform(s)

Linux (x86)

### What version/commit are you on?

1075995e

### Code of Conduct

- [X] I agree to follow the Code of Conduct

**Comment by gakonst** [comment]
https://github.com/erthink/libmdbx/blob/bb8f43181783686879219846d64379a04c1430e3/mdbx.h#L1667-L1668

Yikes!

@Rjected @rakita @mattsse @rkrasiuk this is interesting, so the logs show that we tried to insert a bunch of blocks one after the other via blockchain tree, and then we switch to pipeline at some point and trigger a db corruption (this is weird?)

**Comment by risingsun007** [comment]
I had this issue too "'try_insert_validated_block' loop issue too

**Comment by Rjected** [comment]
This loop is the intended behavior, because we currently download blocks one-by-one, going backwards from the tip, after the pipeline is run for the first time. Our logs for this code is noisy right now but this behavior is expected. The DB error seems more worrying, so I'm not sure whether or not that's due to a bug in reth, or something lower level is responsible.

**Comment by gakonst** [comment]
:+1: Can repro, investigating

**Comment by gakonst** [comment]
Have you pulled / built and retried with https://github.com/paradigmxyz/reth/commit/255780d3810d8e35e012d4bdb0df5b1c3fdf6e9e?

**Comment by gakonst** [comment]
Roger - I pulled + applied the latest main, and now I'm at the tip syncing, so that should resolve the issue, temporarily.

**Comment by mattsse** [comment]
> I don't get the DB corruption but I do get the multiple try_insert_validated_block logs and all APIs go down (I'm running HTTP, engine API, metrics API)

we added an improvement for this in https://github.com/paradigmxyz/reth/pull/3059 which should mitigate most of it.

executing a range of blocks (in the 10s of blocks) at once can currently lead to timed-out CL requests 
however this only happens when reth commits a larger range of blocks for example after restart or after connection loss.

working on improving this further so CL requests don't time out

**Comment by gakonst** [comment]
Closing this optimistically as @garythung has resolved it and I also encountered & #3059 fixed it. @Gaon3 pls re-open if still an issue after pulling latest.
